### ISSUE TITLE
not able to hide the Java variable name while rendering the output

### ISSUE THREAD

**Issue by zappee** [remkop/picocli#1271]
I am using picocli to show the usage of my command-line application. What I am not able to reach is to hide the name of the Java variable which appears in the printed output and from my point of view it looks so ugly.

This is the configutation of my CommandLine.Option:

```
@CommandLine.Option(
        names = {"-a", "--abc"},
        description = "Please, hide the Java variable name...")
private String xxx;

```
And this is how it is rendered:

`-a, --abc=<xxx>   Please, hide the Java variable name...`

As you can see the name of the Java variable appears after the equal sign: `<xxx>`

I would like to hide it, like this:

`-a, --abc   Please, hide the Java variable name...`

I checked the API but I could not see anything related to his. Is there any way to turn it off?

**Comment by remkop** [comment]
You can use the [`paramLabel` attribute](https://picocli.info/#_parameter_labels) to control what is displayed:

```java
@Option(
        names = {"-a", "--abc"},
        paramLabel = "NUMBER"
        description = "Specify any number between 0 and 10...")
private String xxx;

**Comment by zappee** [comment]
I saw that option. Unfortunately it only works if assign a value to the
paramLabel. If I set it to an empty string ("") then the name of the java
variable shows. I would like to see there nothing, just the names and
description. Nothing else.

Is that possible?

Thanks.

**Comment by remkop** [comment]
Okay, have you tried using a space?
Just checking, these are not boolean options, right?
End users _do_ need to specify a value after the option name, correct?

**Comment by zappee** [comment]
Spacee does not work.

These are String and int values.

For boolean variable it works fine: the java variable name does not show.

End user have to give a value.

Example:
-u, --user
Description: The connecting use name.
String user

Or:
-p, --password
Description: Password for the user.
String password

Or:
-o, --port
Description Listening port.
Int port


In the cases above to show variable name is completely useless:

-p, --password=<password>.  Description...

looks not nice.

**Comment by remkop** [comment]
Ok, understood. Thanks for the clarification.

The annotations API does not provide for this, but it is possible to achieve this by overriding the Help API and plugging in your own option renderer. For example:

```java
@Test
public void testIssue1271() {
    class Issue1271 {
        @Option(names = {"-u", "--user"}, description = "The connecting user name.")
        private String user;

        @Option(names = {"-p", "--password"}, interactive = true, description = "Password for the user.")
        private String password;

        @Option(names = {"-o", "--port"}, description = "Listening port.")
        private int port;
    }
    String expected = String.format("" +
            "Usage: <main class> [-p] [-o=<port>] [-u=<user>]%n" +
            "  -o, --port          Listening port.%n" +
            "  -p, --password      Password for the user.%n" +
            "  -u, --user          The connecting user name.%n");
    CommandLine cmd = new CommandLine(new Issue1271());
    cmd.setHelpFactory(new IHelpFactory() {
        public Help create(final CommandSpec commandSpec, ColorScheme colorScheme) {
            return new Help(commandSpec, colorScheme) {
                @Override
                public IOptionRenderer createDefaultOptionRenderer() {
                    return new IOptionRenderer() {
                        public Text[][] render(OptionSpec option, IParamLabelRenderer parameterLabelRenderer, ColorScheme scheme) {
                            String shortOption = option.shortestName(); // assumes every option has a short option
                            String longOption = option.longestName(); // assumes every option has a short and a long option

                            if (option.negatable()) { // ok to omit if you don't have negatable options
                                CommandLine.INegatableOptionTransformer transformer = option.commandSpec.negatableOptionTransformer();
                                shortOption = transformer.makeSynopsis(shortOption, option.commandSpec);
                                longOption = transformer.makeSynopsis(longOption, option.commandSpec);
                            }

                            // assume one line of description text (may contain embedded %n line separators)
                            String[] description = option.description();
                            Text[] descriptionFirstLines = scheme.text(description[0]).splitLines();

                            Text EMPTY = Ansi.EMPTY_TEXT;
                            List<Text[]> result = new ArrayList<Text[]>();
                            result.add(new Text[] {
                                    scheme.optionText(String.valueOf(commandSpec.usageMessage().requiredOptionMarker())),
                                    scheme.optionText(shortOption),
                                    scheme.text(","), // we assume every option has a short and a long name
                                    scheme.optionText(longOption), // just the option name without parameter
                                    descriptionFirstLines[0] });
                            for (int i = 1; i < descriptionFirstLines.length; i++) {
                                result.add(new Text[] { EMPTY, EMPTY, EMPTY, EMPTY, descriptionFirstLines[i] });
                            }
                            return result.toArray(new Text[result.size()][]);
                        }
                    };
                }
            };
        }
    });
    String actual = cmd.getUsageMessage();
    assertEquals(expected, actual);
}
```

(I made password an [interactive option](https://picocli.info/#_interactive_password_options), so users will be prompted when the `-p` option is specified without parameter.)

This prints the below usage help message. Note that the param label is still printed in the synopsis, but omitted in the options list. Does this meet your needs?

```
Usage: <main class> [-p] [-o=<port>] [-u=<user>]
  -o, --port          Listening port.
  -p, --password      Password for the user.
  -u, --user          The connecting user name.
```

**Comment by zappee** [comment]
Thanx for the code. This is exactly what I wanted.

Only two comments, related to your code:

- Text EMPTY = Ansi.**EMPTY_TEXT**
For me, only this was working: `Text EMPTY = Ansi.OFF.new Text(0);`

- I think that you made a typo: **commandSpec** -> `command()`
CommandLine.INegatableOptionTransformer transformer = option.**commandSpec**.negatableOptionTransformer();
shortOption = transformer.makeSynopsis(shortOption, option.**commandSpec**);
longOption = transformer.makeSynopsis(longOption, option.**commandSpec**);


Thanks for your help.

Maybe you can post your solution on [Stackowerflow](https://stackoverflow.com/questions/65130421/picocli-cli-parser-hide-the-java-variable-name-while-rendering-the-output), were also asked for help.

**Comment by zappee** [comment]
What I see now is that the default value is not rendered.

I have set the default values for some params to make end-user's life easier.

When I debug the fix that u have provided I can see that the default value is in the `option.defaultValue`. But I am not sure how to read this value and add properly to the `List<Text[]> result = new ArrayList<Text[]>()`.

Could you please check it?
Thx.

**Comment by zappee** [comment]
I hacked a little bit of your code in order to I can add the wiped default value.
I am sure that my solution is not the best way to do this, but it displays what I need.

Later when you release a new version with the fix, I can remove that hack I made :grin:

**Comment by remkop** [comment]
@zappee Great feedback, thank you!
I will take a more detailed look and post the answer on StackOverflow also, hopefully this weekend.

Meanwhile, I hope you like using picocli. 
(If you do, please star the project on GitHub 😜  and spread the word!)

**Comment by zappee** [comment]
I like picocli so much!
I appreciate the effort that the team invests in this open-source project.
This is my 2nd small project where I use picocli. Both are open source projects, available from [git](https://github.com/zappee?tab=repositories).

Anyway, this is how the help looks like now:
```
Usage: JMS Message Sender [-?v] -s -f=<connectionFactoryJndi> [-h=<webLogicHost>]
                          [-i=<initialContextFactory>] [-p=<webLogicPort>] -q=<queueJndi>
                          [-u=<webLogicUser>] <message>
JMS message sender command line tool. This tool can send messages to the given JMS queue.

General options:
      <message>             The message will be sent to the queue
  -v, --verbose             It provides additional details as to what the tool is doing.
  -i, --icf                 To create a WebLogic context from a client, your code must minimally
                              specify this factory as the initial context factory.
                              Default: weblogic.jndi.WLInitialContextFactory
  -h, --host                The hostname of the machine where the WebLogic server runs.
                              Default: localhost
  -o, --port                The listening T3 port for the WebLogic server.
                              Default: 7001
  -u, --user                The username for the WebLogic server.
                              Default: weblogic
 ...
```

If you have time you can take care of the variable names that appears in the 'header' above.

**Comment by remkop** [comment]
@zappee About default values, I will fix the example, but there is an alternative: 
are you aware you can use the [`${DEFAULT-VALUE}` variable](https://picocli.info/#_predefined_variables) in the option description? For example:

```java
@Option(names = {"-o", "--port"}, description = "Listening port, default is ${DEFAULT-VALUE}.",
        defaultValue = "12345")
private int port;
```

... will show in the usage help as:

```
  -o, --port          Listening port, default is 12345.
```

> If you have time you can take care of the variable names that appears in the 'header' above.

I call this bit of the usage help message the "synopsis":

```
Usage: JMS Message Sender [-?v] -s -f=<connectionFactoryJndi> [-h=<webLogicHost>]
                          [-i=<initialContextFactory>] [-p=<webLogicPort>] -q=<queueJndi>
                          [-u=<webLogicUser>] <message>
```

About the synopsis, I would recommend using descriptive `paramLabel` values that help the end user understand what they are supposed to provide. For example for `-f=<connectionFactoryJndi>`, do users need to provide a JNDI URL, or a factory class name? You could choose a `paramLabel` to help the user, so it could be `-f=JNDI_URL` or `-f=FACTORY_CLASS_FQCN`, or something like that. 
But, if you do not want _any_ parameter labels shown in the synopsis, you can specify a custom synopsis in the `@Command` annotation, see [this section](https://picocli.info/#_custom_synopsis) in the user manual.

Hope that is useful, I will post to StackOverflow later this weekend.

**Comment by remkop** [comment]
I posted an answer to StackOverflow, please take a look.

I also fixed https://github.com/remkop/picocli/issues/1273, to be included in the upcoming picocli 4.6 release.
With this fix, the help message in your example will look a bit nicer: there will be less superfluous space between the option names and the description.
